yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,412 @@
+                f.write(struct.pack('6d4i?2ih',
+                                    self.c_tree.nodes[i].left_edge[0],
+                                    self.c_tree.nodes[i].left_edge[1],
+                                    self.c_tree.nodes[i].left_edge[2],
+                                    self.c_tree.nodes[i].right_edge[0],
+                                    self.c_tree.nodes[i].right_edge[1],
+                                    self.c_tree.nodes[i].right_edge[2],
+                                    self.c_tree.nodes[i].start,
+                                    self.c_tree.nodes[i].end,
+                                    self.c_tree.nodes[i].parent,
+                                    self.c_tree.nodes[i].children,
+                                    self.c_tree.nodes[i].leaf,
+                                    self.c_tree.nodes[i].node_id,
+                                    self.c_tree.nodes[i].leaf_id,
+                                    self.c_tree.nodes[i].depth))
+
+    def load(self, fname = None):
+        if fname is None:
+            raise ValueError("A filename must be specified to load the octtree!")
+        # clear any current tree we have loaded
+        self.reset()
+
+        cdef Node temp
+        with open(fname,'rb') as f:
+            (self.num_particles, self.num_octs, self.n_ref) = \
+                struct.unpack('3i', f.read(12))
+            self.idx = \
+                np.asarray(struct.unpack('{}i'.format(self.num_particles),
+                           f.read(4*self.num_particles)), dtype=np.int32)
+            reserve(&self.c_tree.nodes, self.num_octs+1)
+            for i in range(self.num_octs):
+                (temp.left_edge[0], temp.left_edge[1], temp.left_edge[2],
+                 temp.right_edge[0], temp.right_edge[1], temp.right_edge[2],
+                 temp.start, temp.end, temp.parent, temp.children, temp.leaf,
+                 temp.node_id, temp.leaf_id, temp.depth) = \
+                struct.unpack('6d4i?2ih', f.read(78))
+
+                self.c_tree.nodes.push_back(temp)
+
+    cdef void smooth_onto_leaves(self, np.float64_t[:] buff, np.float64_t posx,
+                                 np.float64_t posy, np.float64_t posz,
+                                 np.float64_t hsml, np.float64_t prefactor,
+                                 Node * node):
+
+        cdef Node * child
+        cdef double q_ij, diff_X, diff_y, diff_z
+
+        if node.leaf == 0:
+            child = &self.c_tree.nodes[node.children]
+            for i in range(8):
+                if child[i].left_edge[0] - posx < hsml and \
+                        posx - child[i].right_edge[0] < hsml:
+                    if child[i].left_edge[1] - posy < hsml and \
+                            posy - child[i].right_edge[1] < hsml:
+                        if child[i].left_edge[2] - posz < hsml and \
+                                posz - child[i].right_edge[2] < hsml:
```

Comment:
You're accessing `child[i]` over and over again. You can do that once before the if statements and avoid the array access overhead. 

Also use parens to avoid using the continuation characters. It might also be worthwhile to add a helper function that does the math and then call that function (maybe make it an inline function for speed), it's a little eyewatering to read long lines like this, especially in tight code.

Reply:
fixed